tests/operator_test.go

```diff
@@ -65,3 +69,103 @@ func TestCRDExplainable(t *testing.T) {
 	Expect(out).To(ContainSubstring("OperatorVersion The version of the HostPathProvisioner Operator"))
 	Expect(out).To(ContainSubstring("TargetVersion The targeted version of the HostPathProvisioner deployment"))
 }
+
+func TestNodeSelector(t *testing.T) {
+	RegisterTestingT(t)
+	tearDown, k8sClient := setupTestCase(t)
+	defer tearDown(t)
+
+	hppClient, err := getHPPClient()
+	Expect(err).ToNot(HaveOccurred())
+
+	cr, err := hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Get(context.TODO(), "hostpath-provisioner", metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	nodes, _ := k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	affinityTestValue := &v1.Affinity{
+		NodeAffinity: &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{Key: "kubernetes.io/hostname", Operator: v1.NodeSelectorOpIn, Values: []string{nodes.Items[0].Name}},
+						},
+					},
+				},
+			},
+		},
+	}
+	nodeSelectorTestValue := map[string]string{"kubernetes.io/arch": "not-a-real-architecture"}
+	tolerationsTestValue := []v1.Toleration{{Key: "test", Value: "123"}}
+
+	origWorkloads := cr.Spec.Workloads.DeepCopy()
+	defer func() {
+		cr, err = hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Get(context.TODO(), "hostpath-provisioner", metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		cr.Spec.Workloads = *origWorkloads.DeepCopy()
+
+		_, err = hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Update(context.TODO(), cr, metav1.UpdateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		err = k8sClient.AppsV1().DaemonSets("hostpath-provisioner").Delete(context.TODO(), "hostpath-provisioner", metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() bool {
+			ds, err := k8sClient.AppsV1().DaemonSets("hostpath-provisioner").Get(context.TODO(), "hostpath-provisioner", metav1.GetOptions{})
+			if err != nil {
+				return false
+			}
+
+			podSpec := ds.Spec.Template.Spec
+
+			if reflect.DeepEqual(podSpec.NodeSelector, nodeSelectorTestValue) {
+				fmt.Printf("Node selector is actually %v\n", podSpec.NodeSelector)
+				return false
+			}
+			return true
+		}, 270*time.Second, 1*time.Second).Should(BeTrue())
+	}()
+
+	cr.Spec.Workloads = hostpathprovisioner.NodePlacement{
+		NodeSelector: nodeSelectorTestValue,
+		Affinity:     affinityTestValue,
+		Tolerations:  tolerationsTestValue,
+	}
+
+	_, err = hppClient.HostpathprovisionerV1beta1().HostPathProvisioners().Update(context.TODO(), cr, metav1.UpdateOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	err = k8sClient.AppsV1().DaemonSets("hostpath-provisioner").Delete(context.TODO(), "hostpath-provisioner", metav1.DeleteOptions{})
```

Comment:
Again this one should not be necessary. once the CR is updated, the operator should update the daemonset and eventually the daemonset should get changed, if it is not then that is a bug.